### ISSUE TITLE
Cannot use simple models derived from str (with Pydantic RootModel) as query params

### ISSUE THREAD

**Issue by Kludex** [tiangolo/fastapi#11134]
### Discussed in https://github.com/tiangolo/fastapi/discussions/11101

<div type='discussions-op-text'>

<sup>Originally posted by **WarpedPixel** February  6, 2024</sup>
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# fail.py
# uvicorn fail:app --reload

from fastapi import FastAPI
from pydantic import RootModel

SimpleID = RootModel[str]

app = FastAPI()

@app.get("/test1")
async def t1(param1: SimpleID):     # BUG: expected to be query param, but it is body
    return {"message": f"Hello {param1}"}

# @app.get("/test2")
# async def t2(param1: Annotated[SimpleID, Query()]): # this workaround does NOT work
#     return {"message": f"Hello {param1}"}

@app.get("/test3")
async def t3(param1: SimpleID, param2: str):  # param2 is query param as expected, param1 is NOT
    return {"message": f"Hello {param1} {param2}"}
```

### Description

Simple types should be usable as query params. Pydantic provides `RootModel` (instead of `BaseModel`) precisely to implement types with rich validation based on native types like `list` or `str` (instead of the more typical object/dict). Types built with `RootModel` do not seem to work as query params, *even if it is functionally equivalent* to a simple type like `str`.
Furthermore, we cannot force them to work as query params with `Query()` annotations either.
**NOTE:** the example is overly simplified, my RootModel-derived types do much more in terms of validation.

### Operating System

macOS

### FastAPI Version

0.108.0

### Pydantic Version

2.5.3

### Python Version

Python 3.11.4

### Additional Context

I am not sure this is possible, but I would like for FastAPI to understand RootModel derived types, if the root type is one of the types that can be used as query param, then it should work as such.
The workaround in my case is to use `str` parameters in my FastAPI calls then manually convert to the rich Pydantic type that does validation. Which kind of defeats the purpose of FastAPI rich type annotations.</div>

**Comment by marianhlavac** [comment]
Stumbled upon this today, I was quite surprised this is not possible, this feels very FastAPI-like 😄 Is this feature difficult to implement?

**Comment by MarkusSintonen** [comment]
@Kludex, @marianhlavac adding support is easy so I did it here: https://github.com/tiangolo/fastapi/pull/11306

Lets see if this gets accepted

**Comment by j-carson** [comment]
These two alternative implementations also work: 

Both alternatives require you to declare the parameter as it's "root" type. 

Validation option 1 constructs the class (forcing the validators to run) then returns just the root portion.
Validation option 2 just validates the input with a simple function.

```python
from fastapi import FastAPI
from pydantic import RootModel, AfterValidator, model_validator
from typing import Annotated

class SimpleID(RootModel[str]):
    # Simulate work to validate the string here
    @model_validator(mode='after')
    @classmethod
    def check_my_string(cls, data: str):
        print(f"Validation of {data} passed!")
        return data

def string_checker(data: str) -> str:
    print(f"Validation of {data} passed!")
    return data


app = FastAPI()

@app.get("/test1")
async def t1(param1: Annotated[str, AfterValidator(lambda v: SimpleID(v).root)]):
    return {"message": f"Hello {param1}"}
    
@app.get("/test2")
async def t1(param1: Annotated[str, AfterValidator(string_checker)]):
    return {"message": f"Hello {param1}"}

```

**Comment by MarkusSintonen** [comment]
> These two alternative implementations also work:

Those don't produce proper OpenAPI specs, or don't allow the Pydantics model Configs?

**Comment by j-carson** [comment]
> > These two alternative implementations also work:
> 
> Those don't produce proper OpenAPI specs, or don't allow the Pydantics model Configs?

Maybe I'm misunderstanding the full requirements here -- What did you want your OpenAPI spec to look like? I see that "param1" is a string in both cases which is what I expected.  You can add Query to the annotation chain to set some more pieces of the OpenAPI spec, if that helps at all:

```python
@app.get("/test1")
async def t1(param1: Annotated[str, Query(description="This param must be a SimpleID"),  AfterValidator(lambda v: SimpleID(v).root)]):
    return {"message": f"Hello {param1}"}
```

If the idea was to have simple scalar root model wrapped around a type that FastAPI already knows how to support -- why would you want/need a config dictionary on that?

**Comment by MarkusSintonen** [comment]
> If the idea was to have simple scalar root model wrapped around a type that FastAPI already knows how to support -- why would you want/need a config dictionary on that?

Idea would be to support more complex validation/schema cases (eg customizing the spec via Config etc). Usually the support for using Pydantic models is great here so why not supporting Pydantic RootModels? The workaround is quite verbose compared to just using a RootModel.

**Comment by j-carson** [comment]
> Usually the support for using Pydantic models is great here so why not supporting Pydantic RootModels? The workaround is quite verbose compared to just using a RootModel.

No, the suppport for using Pydantic models as query parameters is not "great". Everything that you're hitting with RootModel in your original issue - FastAPI wanting to make it a Body rather than a Query, and not accepting a model as a type hint to Query, etc. applies to every type of Pydantic model - not just your RootModel.

**Comment by tiangolo** [comment]
Thanks all for the interest and dicussions here! ☕ 

I have never been a fan of Pydantic's `RootModel`, as it adds impedance on what the type actually is. There was no way to achieve it in Pydantic v1, so I understand its existence. And Pydantic v2 need to have some type of compatibility with v1, so I understand why it's available in Pydantic v2.

I think now that Pydantic v2 supports `AfterValidator` and others, things can be achieved in ways that don't interfere with how types work, without using `RootModel`.

For now I'll pass on this request, but if some people can come up with compelling use cases of why this is needed and can't be achieved with the other regular Pydantic tools (like `AfterValidator`), please create a discussion question with the specific use case and I'll review it again.

For now, I'll close this one. Thanks! ☕
